### PR TITLE
Support for WebExtensions APIs

### PR DESCRIPTION
**Pull request by 149segolte**
Basic Node.js Structure is in place for further work.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,34 @@
+{
+  "name": "motrix-webextension",
+  "version": "2.0.0",
+  "description": "WebExtension for Motrix download manager",
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "build": "webpack --config=webpack.config.js --mode=production",
+    "start:firefox": "npm run build && web-ext run --firefox=firefoxdeveloperedition --source-dir ./dist/",
+    "start:edge": "npm run build && web-ext run -t chromium --chromium-binary edge.exe --source-dir ./dist/"
```
Comment (149segolte) [inline package.json:10]: chromium-based browsers not automatically starting. ( temporarily can load from developer mode. )

File: src/background.js [hunk 2]
```diff
@@ -56,43 +56,52 @@ function downloadAgent() {
                         out: filename
                     };
                 }
-                await aria2.call("addUri", [downloadUrl], params).then(async ()=> {
-                    // Added successfully: Cancels and removes the download from chrome download manager
-                    chrome.downloads.erase({id: downloadItem.id});
+                await aria2.call("addUri", [downloadUrl], params).then(async () => {
+                    // Added successfully: Cancels and removes the download from browser download manager
+                    browser.downloads.erase({ id: downloadItem.id });
```
Comment (149segolte) [inline src/background.js:61]: Downloads.erase() does not seem to work with Firefox. ( Works on chrome/edge )

File: src/manifest.json [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  "background": {
+    "scripts": [
+      "browser-polyfill.min.js",
+      "background.js"
+    ],
+    "persistent": true
+  },
+  "applications": {
+    "gecko": {
+      "id": "gautamkrishnar@github.com",
```
Comment (149segolte) [inline src/manifest.json:38]: Later will require permanent Application ID for Firefox to allow storage API to work.

File: src/background.js (hunk 2 above)
Comment (gautamkrishnar) [reply src/background.js:61]: @149segolte please add these as the comment in the code as well.

File: src/manifest.json [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    "default_popup": "popup.html"
+  },
+  "background": {
+    "scripts": [
+      "browser-polyfill.min.js",
+      "background.js"
+    ],
+    "persistent": true
+  },
+  "browser_specific_settings": {
```
Comment (149segolte) [inline src/manifest.json:36]: @gautamkrishnar Chrome doesn't support "browser_specific_settings", therefore before publishing to a specific browser add and remove this accordingly. I will create a publishing script later.

File: src/manifest.json [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    "gecko": {
+      "id": "gautamkrishnar@github.com",
```
Comment (149segolte) [inline src/manifest.json:38]: @gautamkrishnar Replace this with a permanent ID so that I can test the extension in production environment.

Comment (gautamkrishnar) [reply src/manifest.json:38]: @149segolte addon id is optional: https://extensionworkshop.com/documentation/develop/extensions-and-the-add-on-id/ lets keep it as `r.gautamkrishna@gmail.com`

Comment (149segolte) [reply src/manifest.json:38]: @gautamkrishnar [We require explicit addon ID to use storage.sync API](https://extensionworkshop.com/documentation/develop/extensions-and-the-add-on-id/#when-do-you-need-an-add-on-id:~:text=Some%20WebExtension%20APIs%20use%20the%20add%2Don%20ID%20and%20expect%20it%20to%20be%20the%20same%20from%20one%20browser%20session%20to%20the%20next.%20If%20you%20use%20these%20APIs%2C,runtime.onConnectExternal)

I will update the ID.

File: src/background.js [hunk 6]
```diff
@@ -1,13 +1,17 @@
-'use strict';
+import Aria2 from 'aria2';
+
+function validateUrl(value) {
+    return /^(?:(?:(?:https?|ftp):)?\/\/)(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)(?:\.(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)*(?:\.(?:[a-z\u00a1-\uffff]{2,})))(?::\d{2,5})?(?:[/?#]\S*)?$/i.test(value);
```
Comment (149segolte) [inline src/background.js:4]: @gautamkrishnar can you provide a regex in replacement of this one as it seems to be having some issues.
The regex test can be found at [regex101](https://regex101.com/r/0fPihZ/1).
And the lint report is:
image

File: src/manifest.json (hunk 4 above)
Comment (gautamkrishnar) [reply src/manifest.json:36]: @149segolte it is taken from https://stackoverflow.com/questions/8667070/javascript-regular-expression-to-validate-url

Feel free to use any. I think checking http/https and :// would be enough. You can use any other regex mentioned in the Stackoverflow.

File: src/background.js (hunk 6 above)
Comment (gautamkrishnar) [reply src/background.js:4]: @149segolte it is taken from https://stackoverflow.com/questions/8667070/javascript-regular-expression-to-validate-url

Feel free to use any. I think checking http/https and :// would be enough. You can use any other regex mentioned in the Stackoverflow.

File: package.json [hunk 7]
```diff
@@ -0,0 +1,46 @@
+{
+  "name": "motrix-webextension",
+  "version": "2.0.0",
+  "description": "WebExtension for Motrix download manager",
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "build": "webpack --config=webpack.config.js --mode=production",
+    "start:firefox": "npm run build && web-ext run --firefox=firefoxdeveloperedition --source-dir ./dist/",
+    "package": "npm run build && node package.js chrome firefox"
```
Comment (149segolte) [inline package.json:10]: @gautamkrishnar The package script seems to throw an error but executes perfectly. So ignore the error while executing it. I think most of the things are in place to test and publish the extension for chrome, edge and firefox.

File: src/background.js [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+                        type: 'basic',
+                        iconUrl: 'assets/images/icon-large.png',
+                        title: 'Motrix WebExtension',
+                        message: 'Download started in Motrix download manger'
+                    };
+                    var notificationId = Math.round((new Date()).getTime() / 1000).toString();
+                    browser.notifications.create(notificationId, notificationOptions);
+                    browser.notifications.onClicked.addListener((id) => {
+                        if (id == notificationId) {
+                            browser.tabs.create({url:'motrix://'});
```
Comment (149segolte) [inline src/background.js:81]: This does seem to work on firefox but not on chrome, help required.

File: package.js [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+args.forEach(function (val) {
+    var zip = val + '.zip';
+    var zipPath = path.resolve(packagePath, zip);
+    console.log('\n\'' + val + '\' target detected ...');
+    if (val == 'chrome') {
+        console.log('Removing Browser Specific Settings from the manifest ...');
+        delete file['browser_specific_settings'];
+        fs.writeFileSync(filePath, JSON.stringify(file), function writeJSON(err) {
```
Comment (149segolte) [inline package.js:56]: @gautamkrishnar The script doesn't seem to be able to remove the browser-specific settings. Please look into this and provide some feedback. Thanks in advance.

**Comment by 149segolte** [comment]
The only issues remaining are,

1. The package script seems to archive after the manifest is restored.
2. The event onClicked of the notification does not seem to trigger motrix on chrome and edge ( works fine on firefox ).

**Comment by 149segolte** [comment]
@gautamkrishnar This branch is tested on Microsoft Edge Stable and Firefox Developer Edition. Please provide testing results from Chrome.
Opera and other chromium-based browsers should work but are not explicitly tested.

**Comment by gautamkrishnar** [comment]
@149segolte thanks a lot, I will give it a go this weekend. 👍 ⭐

**Comment by gautamkrishnar** [comment]
@149segolte thanks a lot for the PR, you rocks. Will make some changes and add a CI so that users don't have to follow all the build step by themselves. Thanks a lot for taking this up.

**Pull request merged**